File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/operations/RedshiftS3StagingSqlOperations.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.redshift.operations;

import com.amazonaws.services.s3.AmazonS3;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.airbyte.cdk.db.jdbc.JdbcDatabase;
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer;
import io.airbyte.cdk.integrations.destination.record_buffer.SerializableBuffer;
import io.airbyte.cdk.integrations.destination.s3.AesCbcEnvelopeEncryption;
import io.airbyte.cdk.integrations.destination.s3.AesCbcEnvelopeEncryptionBlobDecorator;
import io.airbyte.cdk.integrations.destination.s3.EncryptionConfig;
import io.airbyte.cdk.integrations.destination.s3.S3DestinationConfig;
import io.airbyte.cdk.integrations.destination.s3.S3StorageOperations;
import io.airbyte.cdk.integrations.destination.s3.credential.S3AccessKeyCredentialConfig;
import io.airbyte.cdk.integrations.destination.staging.StagingOperations;
import io.airbyte.commons.lang.Exceptions;
import io.airbyte.integrations.destination.redshift.manifest.Entry;
import io.airbyte.integrations.destination.redshift.manifest.Manifest;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Base64;
import java.util.Base64.Encoder;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RedshiftS3StagingSqlOperations extends RedshiftSqlOperations implements StagingOperations {

  private static final Encoder BASE64_ENCODER = Base64.getEncoder();
  private final NamingConventionTransformer nameTransformer;
  private final S3StorageOperations s3StorageOperations;
  private final S3DestinationConfig s3Config;
  private final ObjectMapper objectMapper;
  private final byte[] keyEncryptingKey;

  private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftS3StagingSqlOperations.class);

  public RedshiftS3StagingSqlOperations(final NamingConventionTransformer nameTransformer,
                                        final AmazonS3 s3Client,
                                        final S3DestinationConfig s3Config,
                                        final EncryptionConfig encryptionConfig) {
    this.nameTransformer = nameTransformer;
    this.s3StorageOperations = new S3StorageOperations(nameTransformer, s3Client, s3Config);
    this.s3Config = s3Config;
    this.objectMapper = new ObjectMapper();
    if (encryptionConfig instanceof final AesCbcEnvelopeEncryption e) {
      this.s3StorageOperations.addBlobDecorator(new AesCbcEnvelopeEncryptionBlobDecorator(e.key()));
      this.keyEncryptingKey = e.key();
    } else {
      this.keyEncryptingKey = null;
    }
  }

  @Override
  public String getStagingPath(final UUID connectionId,
                               final String namespace,
                               final String streamName,
                               final String outputTableName,
                               final Instant writeDatetime) {
    final String bucketPath = s3Config.getBucketPath();
    final String prefix = bucketPath.isEmpty() ? "" : bucketPath + (bucketPath.endsWith("/") ? "" : "/");
    final ZonedDateTime zdt = writeDatetime.atZone(ZoneOffset.UTC);
    return nameTransformer.applyDefaultCase(String.format("%s%s/%s_%02d_%02d_%02d_%s/",
        prefix,
        nameTransformer.applyDefaultCase(nameTransformer.convertStreamName(outputTableName)),
        zdt.getYear(),
        zdt.getMonthValue(),
        zdt.getDayOfMonth(),
        zdt.getHour(),
        connectionId));
  }

  @Override
  public String getStageName(final String namespace, final String streamName) {
    return "garbage-unused";
  }

  @Override
  public void createStageIfNotExists(final JdbcDatabase database, final String stageName) throws Exception {
    s3StorageOperations.createBucketIfNotExists();
  }

  @Override
  public String uploadRecordsToStage(final JdbcDatabase database,
                                     final SerializableBuffer recordsData,
                                     final String schemaName,
                                     final String stageName,
                                     final String stagingPath)
      throws Exception {
    return s3StorageOperations.uploadRecordsToBucket(recordsData, schemaName, stagingPath);
  }

  private String putManifest(final String manifestContents, final String stagingPath) {
    final String manifestFilePath = stagingPath + String.format("%s.manifest", UUID.randomUUID());
    s3StorageOperations.uploadManifest(manifestFilePath, manifestContents);
    return manifestFilePath;
  }

  @Override
  public void copyIntoTableFromStage(final JdbcDatabase database,
                                     final String stageName,
                                     final String stagingPath,
                                     final List<String> stagedFiles,
                                     final String tableName,
                                     final String schemaName)
      throws Exception {
    LOGGER.info("Starting copy to target table from stage: {} in destination from stage: {}, schema: {}, .",
        tableName, stagingPath, schemaName);
    final var possibleManifest = Optional.ofNullable(createManifest(stagedFiles, stagingPath));
    Exceptions.toRuntime(() -> possibleManifest.stream()
        .map(manifestContent -> putManifest(manifestContent, stagingPath))
        .forEach(manifestPath -> executeCopy(manifestPath, database, schemaName, tableName)));
    LOGGER.info("Copy to target table {}.{} in destination complete.", schemaName, tableName);
  }

  /**
   * Generates the COPY data from staging files into target table
   */
  private void executeCopy(final String manifestPath, final JdbcDatabase db, final String schemaName, final String tableName) {
    final S3AccessKeyCredentialConfig credentialConfig = (S3AccessKeyCredentialConfig) s3Config.getS3CredentialConfig();
    final String encryptionClause;
    if (keyEncryptingKey == null) {
      encryptionClause = "";
    } else {
      encryptionClause = String.format(" encryption = (type = 'aws_cse' master_key = '%s')", BASE64_ENCODER.encodeToString(keyEncryptingKey));
    }

    final var copyQuery = String.format(
        """
        COPY %s.%s FROM '%s'
        CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'
        %s
        CSV GZIP
        REGION '%s' TIMEFORMAT 'auto'
        STATUPDATE OFF
        MANIFEST;""",
        schemaName,
        tableName,
        getFullS3Path(s3Config.getBucketName(), manifestPath),
        credentialConfig.getAccessKeyId(),
        credentialConfig.getSecretAccessKey(),
        encryptionClause,
        s3Config.getBucketRegion());

    Exceptions.toRuntime(() -> db.execute(copyQuery));
  }

  private String createManifest(final List<String> stagedFiles, final String stagingPath) {
    if (stagedFiles.isEmpty()) {
      return null;
    }

    final var s3FileEntries = stagedFiles.stream()
        .map(file -> new Entry(getManifestPath(s3Config.getBucketName(), file, stagingPath)))
        .collect(Collectors.toList());
    final var manifest = new Manifest(s3FileEntries);

    return Exceptions.toRuntime(() -> objectMapper.writeValueAsString(manifest));
  }

  private static String getFullS3Path(final String s3BucketName, final String s3StagingFile) {
    return String.join("/", "s3:/", s3BucketName, s3StagingFile);
  }

  private static String getManifestPath(final String s3BucketName, final String s3StagingFile, final String stagingPath) {
    return "s3://" + s3BucketName + "/" + stagingPath + s3StagingFile;
  }

  @Override
  public void dropStageIfExists(final JdbcDatabase database, final String stageName, final String stagingPath) throws Exception {
    // stageName is unused here but used in Snowflake. This interface needs to be fixed.
    s3StorageOperations.dropBucketObject(stagingPath);
  }

}

File: airbyte-integrations/connectors/destination-redshift/src/main/kotlin/io/airbyte/integrations/destination/redshift/operations/RedshiftS3StagingSqlOperations.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.integrations.destination.redshift.operations

import com.amazonaws.services.s3.AmazonS3
import com.fasterxml.jackson.databind.ObjectMapper
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import io.airbyte.cdk.db.jdbc.JdbcDatabase
import io.airbyte.cdk.integrations.destination.NamingConventionTransformer
import io.airbyte.cdk.integrations.destination.record_buffer.SerializableBuffer
import io.airbyte.cdk.integrations.destination.s3.AesCbcEnvelopeEncryption
import io.airbyte.cdk.integrations.destination.s3.AesCbcEnvelopeEncryptionBlobDecorator
import io.airbyte.cdk.integrations.destination.s3.EncryptionConfig
import io.airbyte.cdk.integrations.destination.s3.S3DestinationConfig
import io.airbyte.cdk.integrations.destination.s3.S3StorageOperations
import io.airbyte.cdk.integrations.destination.s3.credential.S3AccessKeyCredentialConfig
import io.airbyte.cdk.integrations.destination.staging.StagingOperations
import io.airbyte.commons.lang.Exceptions.toRuntime
import io.airbyte.integrations.destination.redshift.manifest.Entry
import io.airbyte.integrations.destination.redshift.manifest.Manifest
import java.time.Instant
import java.time.ZoneOffset
import java.util.*
import java.util.stream.Collectors
import org.slf4j.Logger
import org.slf4j.LoggerFactory

@SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
class RedshiftS3StagingSqlOperations(
    private val nameTransformer: NamingConventionTransformer,
    s3Client: AmazonS3?,
    private val s3Config: S3DestinationConfig,
    encryptionConfig: EncryptionConfig
) : RedshiftSqlOperations(), StagingOperations {
    private val s3StorageOperations = S3StorageOperations(nameTransformer, s3Client!!, s3Config)
    private val objectMapper = ObjectMapper()
    private val keyEncryptingKey: ByteArray?

    init {
        if (encryptionConfig is AesCbcEnvelopeEncryption) {
            s3StorageOperations.addBlobDecorator(
                AesCbcEnvelopeEncryptionBlobDecorator(encryptionConfig.key)
            )
            this.keyEncryptingKey = encryptionConfig.key
        } else {
            this.keyEncryptingKey = null
        }
    }

    override fun getStagingPath(
        connectionId: UUID?,
        namespace: String?,
        streamName: String?,
        outputTableName: String?,
        writeDatetime: Instant?
    ): String? {
        val bucketPath = s3Config.bucketPath
        val prefix =
            if (bucketPath!!.isEmpty()) ""
            else bucketPath + (if (bucketPath.endsWith("/")) "" else "/")
        val zdt = writeDatetime!!.atZone(ZoneOffset.UTC)
        return nameTransformer.applyDefaultCase(
            String.format(
                "%s%s/%s_%02d_%02d_%02d_%s/",
                prefix,
                nameTransformer.applyDefaultCase(
                    nameTransformer.convertStreamName(outputTableName!!)
                ),
                zdt.year,
                zdt.monthValue,
                zdt.dayOfMonth,
                zdt.hour,
                connectionId
            )
        )
    }

    override fun getStageName(namespace: String?, streamName: String?): String? {
        return "garbage-unused"
    }

    @Throws(Exception::class)
    override fun createStageIfNotExists(database: JdbcDatabase?, stageName: String?) {
        s3StorageOperations.createBucketIfNotExists()
    }

    @Throws(Exception::class)
    override fun uploadRecordsToStage(
        database: JdbcDatabase?,
        recordsData: SerializableBuffer?,
        schemaName: String?,
        stageName: String?,
        stagingPath: String?
    ): String {
        return s3StorageOperations.uploadRecordsToBucket(recordsData!!, schemaName, stagingPath!!)
    }

    private fun putManifest(manifestContents: String, stagingPath: String?): String {
        val manifestFilePath = stagingPath + String.format("%s.manifest", UUID.randomUUID())
        s3StorageOperations.uploadManifest(manifestFilePath, manifestContents)
        return manifestFilePath
    }

    @Throws(Exception::class)
    override fun copyIntoTableFromStage(
        database: JdbcDatabase?,
        stageName: String?,
        stagingPath: String?,
        stagedFiles: List<String>?,
        tableName: String?,
        schemaName: String?
    ) {
        LOGGER.info(
            "Starting copy to target table from stage: {} in destination from stage: {}, schema: {}, .",
            tableName,
            stagingPath,
            schemaName
        )
        val possibleManifest = Optional.ofNullable(createManifest(stagedFiles, stagingPath))
        toRuntime {
            possibleManifest
                .stream()
                .map { manifestContent: String -> putManifest(manifestContent, stagingPath) }
                .forEach { manifestPath: String ->
                    executeCopy(manifestPath, database, schemaName, tableName)
                }
        }
        LOGGER.info("Copy to target table {}.{} in destination complete.", schemaName, tableName)
    }

    /** Generates the COPY data from staging files into target table */
    private fun executeCopy(
        manifestPath: String,
        db: JdbcDatabase?,
        schemaName: String?,
        tableName: String?
    ) {
        val credentialConfig = s3Config.s3CredentialConfig as S3AccessKeyCredentialConfig?
        val encryptionClause =
            if (keyEncryptingKey == null) {
                ""
            } else {
                String.format(
                    " encryption = (type = 'aws_cse' master_key = '%s')",
                    BASE64_ENCODER.encodeToString(keyEncryptingKey)
                )
            }

        val copyQuery =
            String.format(
                """
        COPY %s.%s FROM '%s'
        CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'
        %s
        CSV GZIP
        REGION '%s' TIMEFORMAT 'auto'
        STATUPDATE OFF
        MANIFEST;
        """.trimIndent(),
                schemaName,
                tableName,
                getFullS3Path(s3Config.bucketName, manifestPath),
                credentialConfig!!.accessKeyId,
                credentialConfig.secretAccessKey,
                encryptionClause,
                s3Config.bucketRegion
            )

        toRuntime { db!!.execute(copyQuery) }
    }

    private fun createManifest(stagedFiles: List<String?>?, stagingPath: String?): String? {
        if (stagedFiles!!.isEmpty()) {
            return null
        }

        val s3FileEntries =
            stagedFiles
                .stream()
                .map { file: String? ->
                    Entry(getManifestPath(s3Config.bucketName, file!!, stagingPath))
                }
                .collect(Collectors.toList())
        val manifest = Manifest(s3FileEntries)

        return toRuntime<String> { objectMapper.writeValueAsString(manifest) }
    }

    @Throws(Exception::class)
    override fun dropStageIfExists(
        database: JdbcDatabase?,
        stageName: String?,
        stagingPath: String?
    ) {
        // stageName is unused here but used in Snowflake. This interface needs to be fixed.
        s3StorageOperations.dropBucketObject(stagingPath!!)
    }

    companion object {
        private val BASE64_ENCODER: Base64.Encoder = Base64.getEncoder()
        private val LOGGER: Logger =
            LoggerFactory.getLogger(RedshiftS3StagingSqlOperations::class.java)

        private fun getFullS3Path(s3BucketName: String?, s3StagingFile: String): String {
            return java.lang.String.join("/", "s3:/", s3BucketName, s3StagingFile)
        }

        private fun getManifestPath(
            s3BucketName: String?,
            s3StagingFile: String,
            stagingPath: String?
        ): String {
            return "s3://$s3BucketName/$stagingPath$s3StagingFile"
        }
    }
}
